processor/src/range/mod.rs

```diff
@@ -356,13 +275,59 @@ fn lookups_to_rows(num_lookups: usize) -> usize {
     }
 }
 
-/// Returns the number of trace rows needed to describe the specified 8-bit lookup table.
-fn get_num_8bit_rows(lookups: &[usize; 256]) -> usize {
-    let mut result = 0;
-    for &num_lookups in lookups.iter() {
-        result += lookups_to_rows(num_lookups);
+/// Calculates the number of bridge rows that are need to be added to the trace between two values
+/// to be range checked.
+pub fn get_num_bridge_rows(delta: u16) -> usize {
+    let mut exp = 7;
+    let mut gap = delta;
+    let mut bridge_rows = 0_usize;
+    let mut stride = 3_u16.pow(exp);
+    while gap > 0 {
+        if gap >= stride {
+            bridge_rows += 1;
+            gap -= stride;
+        } else {
+            exp = exp.saturating_sub(1);
+            stride = 3_u16.pow(exp);
+        }
+    }
+    bridge_rows.saturating_sub(1)
+}
+
+/// Adds a row for the values to be range checked. In case the difference between the current and
+/// next value is not a power of 3, this function will add additional bridge rows to the trace.
+fn write_rows(
+    trace: &mut [Vec<Felt>],
+    step: &mut usize,
+    num_lookups: usize,
+    value: u16,
+    prev_value: u16,
+    row_flags: &mut [RangeCheckFlag],
+) {
+    let mut exp = 7;
+    let mut gap = value - prev_value;
+    let mut prev_val = prev_value;
+    let mut stride = 3_u16.pow(exp);
+    if gap == 0 {
+        write_value(trace, step, num_lookups, value as u64, row_flags);
+    }
+    while gap > 0 {
+        match gap.cmp(&stride) {
+            Ordering::Equal => {
+                gap -= stride;
+                write_value(trace, step, num_lookups, value as u64, row_flags);
+            }
+            Ordering::Greater => {
+                gap -= stride;
+                prev_val += stride;
+                write_value(trace, step, 0, prev_val as u64, row_flags);
+            }
+            Ordering::Less => {
+                exp = exp.saturating_sub(1);
+                stride = 3_u16.pow(exp);
+            }
+        }
```

Comment:
If I understood the algorithm correctly, we should end up in the `Ordering::Equal` arm at most once, right? Maybe we should put an assertion for this? Or maybe we can simplify this block as something like:
```Rust
while gap != stride {
    if gap > stride {
        gap -= stride;
        prev_val += stride;
        write_value(trace, step, 0, prev_val as u64, row_flags);
    } else {
        stride /= 3;
    }
}
write_value(trace, step, num_lookups, value as u64, row_flags);
```

Reply:
Thanks, done.